src/securityPolicy.c

```diff
@@ -146,53 +145,111 @@ security_policy_t policyForReturnDeriveAddress(const addressParams_t* addressPar
 {
 	DENY_UNLESS(isValidAddressParams(addressParams));
 
-	switch (bip44_classifyPath(&addressParams->spendingKeyPath)) {
+	switch (addressParams->type) {
 
-	case PATH_WALLET_SPENDING_KEY:
-	case PATH_WALLET_STAKING_KEY:
-		if (bip44_isPathReasonable(&addressParams->spendingKeyPath)) {
-			PROMPT();
-		} else {
-			WARN();
-		}
+	case BASE_PAYMENT_KEY_STAKE_KEY:
+		DENY_IF(bip44_classifyPath(&addressParams->spendingKeyPath) != PATH_ORDINARY_SPENDING_KEY);
+		DENY_IF(
+		        addressParams->stakingDataSource == STAKING_KEY_PATH &&
+		        bip44_classifyPath(&addressParams->stakingKeyPath) != PATH_ORDINARY_STAKING_KEY
+		);
+
+		WARN_IF(!bip44_isPathReasonable(&addressParams->spendingKeyPath));
+		WARN_IF(
+		        addressParams->stakingDataSource == STAKING_KEY_PATH &&
+		        !bip44_isPathReasonable(&addressParams->stakingKeyPath)
+		);
+		break;
+
+	case BASE_PAYMENT_KEY_STAKE_SCRIPT:
+	case POINTER_KEY:
+	case ENTERPRISE_KEY:
+	case BYRON:
+		DENY_IF(bip44_classifyPath(&addressParams->spendingKeyPath) != PATH_ORDINARY_SPENDING_KEY);
+
+		WARN_IF(!bip44_isPathReasonable(&addressParams->spendingKeyPath));
+		break;
+
+	case BASE_PAYMENT_SCRIPT_STAKE_KEY:
+	case REWARD_KEY:
+		DENY_IF(addressParams->stakingDataSource != STAKING_KEY_PATH);
+		DENY_IF(bip44_classifyPath(&addressParams->stakingKeyPath) != PATH_ORDINARY_STAKING_KEY);
+
+		WARN_IF(!bip44_isPathReasonable(&addressParams->stakingKeyPath));
+		break;
+
+	case BASE_PAYMENT_SCRIPT_STAKE_SCRIPT:
+	case POINTER_SCRIPT:
+	case ENTERPRISE_SCRIPT:
+	case REWARD_SCRIPT:
+		// no paths in the address
 		break;
 
 	default:
 		DENY();
 		break;
 	}
 
-	DENY(); // should not be reached
+	PROMPT();
 }
 
 // Derive address and show it to the user
 security_policy_t policyForShowDeriveAddress(const addressParams_t* addressParams)
 {
 	DENY_UNLESS(isValidAddressParams(addressParams));
 
-	switch (bip44_classifyPath(&addressParams->spendingKeyPath)) {
+	switch (addressParams->type) {
```

Comment:
Do `policyForReturnDeriveAddress` and `policyForShowDeriveAddress` need to be separated as they are the same except for the last `SHOW()`/`PROMPT()`? Wouldn't it make sense to extract the common part into for example `policyForDeriveAddress`?

Reply:
Yes, saved 190 bytes of code :)

Reply:
Hah, nice 👍